PageObjects/ShopPage.cs

```diff
@@ -1,43 +1,139 @@
+            var addToCartBtnLocator = "//*[@class[starts-with(., 'add-product-to-cart')]]";
+            var addPopup = page.Locator("//*[@id='cart-modal']");
+
+            // Get all product holders on the page
+            var productCards = await page.QuerySelectorAllAsync(productHolderLocator);
+            Assert.That(productCards, Is.Not.Empty, "Products not found on the page");
+
+            // Iterate through each product holder to find the {addProduct}
+            bool isProductFound = false; // To track if product is found
+            foreach (var productCard in productCards)
+            {
+                var productName = await productCard.InnerTextAsync();
+                if (productName.ToLower().Contains(addProduct.ToLower()))
+                {
+                    isProductFound = true;
+                    var addToCartButton = await productCard.QuerySelectorAsync(addToCartBtnLocator);
+                    Assert.That(addToCartButton, Is.Not.Null, $"'Add to cart button' not found for product {addProduct}");
+
+                    //Add specified product to the cart
+                    if (addToCartButton != null)
+                    {
+                        await addToCartButton.ClickAsync();
+                    }
+                    await Assertions.Expect(addPopup).ToContainTextAsync($"{addProduct}");
+                    await Assertions.Expect(addPopup).ToContainTextAsync($"Товар додано у кошик");
+
+                    return;
+                }
+            }
+            Assert.That(isProductFound, Is.True, $"Product '{addProduct}' is not found on the page");
         }
 
-        public async Task IsPageH1Visible(string heading)
+        public async Task ClickLinkButton(string buttonName)
         {
-            await Assertions.Expect(Page.GetByRole(AriaRole.Heading, new() { Name = heading })).ToBeVisibleAsync();
+            var button = page.GetByRole(AriaRole.Link, new() { Name = $"{buttonName}" });
+            await button.ClickAsync();
+        }
+
+        public async Task ClickSpecifiedProductHolder(string productName)
+        {
+            var productHolderLocator = "//*[contains(@class, 'card z-depth-1 hoverable')]";
+            var products = await page.QuerySelectorAllAsync(productHolderLocator);
+            Assert.That(products, Is.Not.Empty, "Products not found on the page");
+
+            foreach (var product in products)
+            {
+                var productText = await product.InnerTextAsync();
+                if (productText.ToLower().Contains(productName.ToLower()))
+                {
+                    await product.ClickAsync();
+                    return;
+                }
+            }
+            await Assertions.Expect(page.Locator("//*[contains(@class, 'availability')]")).ToBeVisibleAsync();
+        }
+
+        public async Task VerifyAddPopupNotVisible()
+        {
+            var addModal = page.Locator("//*[@id='cart-modal']");
```

Comment:
do not use locator with only one assertion .Not.ToBeVisibleAsync();
In case locator changes, this step will be always green. You can make
private Locator addProductPopup = page.Locator("//*[@id='cart-modal']");
and then use it in 2 methods like 
public async Task VerifyAddPopupIsVisible()
public async Task VerifyAddPopupIsNotVisible()